### PR TITLE
qmidiarp: init at 0.6.5

### PR DESCRIPTION
**Pull request by sjfloat**
###### Motivation for this change

Introduction of qmidiarp app.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"` *New pkg, no dependents.*
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/audio/qmidiarp/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
+{ stdenv
+, fetchgit
+, automake
+, autoreconfHook
+, lv2
+, pkg-config
+, qt5
```
Comment (symphorien) [inline pkgs/applications/audio/qmidiarp/default.nix]: You should not take qt5 in argument, use libsForQt5.callPackage instead: see https://nixos.org/nixpkgs/manual/#sec-language-qt

File: pkgs/applications/audio/qmidiarp/default.nix [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  buildInputs = [
+    alsaLib
+    lv2
+    libjack2
+  ] ++ (with qt5; [
+    qttools
+  ]);
+
+  meta = with stdenv.lib; {
+    description = "An advanced MIDI arpeggiator, programmable step sequencer and LFO for Linux. It can hold any number of arpeggiator, sequencer, or LFO modules running in parallel.";
```
Comment (symphorien) [inline pkgs/applications/audio/qmidiarp/default.nix]: maybe split into `description` and `longDescription` ?

**Review by symphorien — CHANGES_REQUESTED** [review]
Thanks for your PR. I left some comments to improve it a bit.

File: pkgs/applications/audio/qmidiarp/default.nix (hunk 2 above)
Comment (sjfloat) [reply pkgs/applications/audio/qmidiarp/default.nix]: Sounds good.

File: pkgs/applications/audio/qmidiarp/default.nix (hunk 1 above)
Comment (sjfloat) [reply pkgs/applications/audio/qmidiarp/default.nix]: I'm a little afraid of this becoming circular as I'm following advise from comments from an earlier PR.

Comment (sjfloat) [reply pkgs/applications/audio/qmidiarp/default.nix]: Is there a reason that one way would be preferable to the other?

**Comment by sjfloat** [comment]
Do the checks reveal that I need to correct the sha256? It builds for me OK locally, but I'm not sure it aligns with the rev being used. Should it?

**Comment by symphorien** [comment]
Try to replace the hash to all zeroes, and then rebuild. You should get the same good hash than in the error message.

File: pkgs/applications/audio/qmidiarp/default.nix (hunk 1 above)
Comment (symphorien) [reply pkgs/applications/audio/qmidiarp/default.nix]: I am not sure about the specifics, but this has something to do with preventing you from pulling dependencies built with different versions of qt by mistake.
We can wait for the review of someone more knowledgeable than me in qt things.

Comment (sjfloat) [reply pkgs/applications/audio/qmidiarp/default.nix]: Sure. Thanks very much for looking at this!

**Comment by sjfloat** [comment]
> Try to replace the hash to all zeroes, and then rebuild. You should get the same good hash than in the error message.

Nice tip. `nix-prefetch-git` accepts a `--rev` argument, so I think I got the right one now. Note that I'm not getting an error message with either one building locally. But one of the @ofborg tests gave errors (though it passed everthing).

**Comment by symphorien** [comment]
Yes, if you have downloaded something (possibly unrelated) with this hash in the past, nix won't even try to download it again, even if you change the url.

**Comment by sjfloat** [comment]
BTW, I'm still not able to request reviews -- it's greyed out. I could ping @jonringer or @jtojnar, but I don't want to overburden them. Is there a more general way?

**Comment by symphorien** [comment]
Huh I don't really know the inner workings or github...

File: pkgs/applications/audio/qmidiarp/default.nix (hunk 1 above)
Comment (jtojnar) [reply pkgs/applications/audio/qmidiarp/default.nix]: I prefer just using the default `mkDerivation` and `callPackage`. As long as you access all Qt dependencies through the same argument, decoherence should not occur. Using Qt’s own functions is not really composable, see https://github.com/NixOS/nixpkgs/pull/71089.

**Comment by jtojnar** [comment]
> Do the checks reveal that I need to correct the sha256? It builds for me OK locally, but I'm not sure it aligns with the rev being used. Should it?

Using `nix-build -A foo.src --check` might also work.

File: pkgs/applications/audio/qmidiarp/default.nix (hunk 1 above)
Comment (sjfloat) [reply pkgs/applications/audio/qmidiarp/default.nix]: Thanks

**Comment by sjfloat** [comment]
> > Do the checks reveal that I need to correct the sha256? It builds for me OK locally, but I'm not sure it aligns with the rev being used. Should it?
> 
> Using `nix-build -A foo.src --check` might also work.

Nice. I need to remember that!

I'll push up with the corrected hash.

**Comment by sjfloat** [comment]
I checked the data files produced by the app before and after adjusting the hash. It was indeed pulling from master. It's correct now.

**Review by symphorien — APPROVED** [review]
* Compiles locally
* Runs fine, even in an empty PATH

**Comment by jonringer** [comment]
> Nice. I need to remember that!

You can also just put in an invalid sha (I usually just change the last character to a). 

This may seem like a hack (it is), but some fetch helpers such as `fetchpatch` will do some processing on the contents and change the contents before determining the sha, and if you do a `nix-prefetch`, you'll get cached hits (the sha before fetchpatch does its processing).

**Review by jonringer — APPROVED** [review]
diff LGTM
shows usage

```
[3 built, 1 copied (1.1 MiB), 0.2 MiB DL]
https://github.com/NixOS/nixpkgs/pull/82795
1 package built:
qmidiarp
```

**Comment by jonringer** [comment]
@GrahamcOfBorg build qmidiarp

**Comment by sjfloat** [comment]
@jonringer is there anything holding this up at this point?

**Comment by jonringer** [comment]
no, I was just busy, sorry

**Comment by sjfloat** [comment]
> no, I was just busy, sorry

No apologies! You're doing a great job!

Thanks

**Pull request merged**